Home Work № 4

skyseedim opened the pull request.

skyseedim committed:
Home Work № 4

NicknameSoHard commented on sal_count.py:
```diff
@@ -0,0 +1,10 @@
+sal_count, vir, stav, bonus = argv
+def salary(x, y, z):
+    try:
+        return(float(x) * float(y) + float(z))
+    except ValueError: # если премия в процентах
+        p = float(z.strip('%')) / 100
+        return (float(x) * float(y) * (1 + p))
```
Чтобы дважды одно и то же не писать, проверили бы тогда нормально ли преобразуются все переменные, а затем уже просто посчитали. Будет чуть лучше)

NicknameSoHard commented on HW4.py:
```diff
@@ -0,0 +1,106 @@
+print("\nTask № 4")
+my_list = [2, 2, 2, 7, 23, 1, 44, 44, 3, 2, 10, 7, 4, 11]
+print("Исходный список: \n", my_list)
+res = [el for el in my_list if my_list.count(el) == 1]
```
В принципе все верно, если интересно я бы предложил подумать над более эффективным решением. метод `count `при каждом вызове перебирает всю коллекцию чтобы проверить кол-во повтором элементов. Т.е. для каждого элементы вы каждый раз пробегаетесь по всей коллекции чтобы его подсчитать, даже если вы уже этот элемент считали.
Более эффективно наверное за 1 проход посчитать все элементы, а затем на второй уже исключить ненужные. Для такого отлично подойдут словари, например. Или `Counter`, о котором на вебинаре упоминали.

NicknameSoHard commented on HW4.py:
```diff
@@ -0,0 +1,106 @@
+
+print("\nTask № 5")
+from functools import reduce
+my_list = [el for el in range(100, 1001) if el % 2 == 0]
```
В `range `можно аргументом указать еще шаг, так что можно сразу сгенерировать без лишних элементов и проверок на это.

NicknameSoHard commented on HW4.py:
```diff
@@ -0,0 +1,106 @@
+print("\nTask № 7")
+from math import factorial
+def fact(n):
+    for i in range(1, n+1):
+        yield factorial(i)
```
Как вариант так, но опять же, если реализовать ручками, то можно более эффективную реализацию сделать. Так как эта функция для каждого числа заново считает все произведение, не используя то, что на прошлых итерациях мы уже все нужное кроме одного умножения посчитали)
Если интересно, распишите сам алгоритм. Там две строчки добавить буквально.

А еще по идее условие остановки нужно передавать аргументом в генератор, а не останавливать во внешнем цикле. Как-раз так будет удобнее сделать при более эффективной реализации.

skyseedim committed:
Изменения в решения задач 4, 5, 7

Задача 4:  Теперь для подсчёта используется Counter, а не метод count, т.е.  теперь для каждого элемента не перебираем всю коллекцию, чтобы  его подсчитать

Задача 5:  Теперь в range добавлен шаг (= 2) и, соответственно, исключена проверка на чётность с помощью if.

Задача 7:  Теперь для подсчёта факториала, не используется ф. factorial их math, а просто умножаются числа друг на друга в цикле for , таким образом, функция для каждого числа заново не считает все произведение с каждой итерацией цикла, а использует то, что на прошлых итерациях мы уже все нужное кроме одного умножения посчитали.
